### PR TITLE
Added test websocket events

### PR DESCRIPTION
**Pull request by jasperpeeters**
Here is the big one about the websocket events. I had to modify the original source from our reg-test-suite quite a lot, I hope took everything into account..

Improvements and comments are welcome.

### CODE REVIEW

File: IS0402Test.py [hunk 1]
```diff
@@ -816,6 +819,236 @@ def test_30(self):
...
+            for k, v in websockets.items():
+                v.close()
+
+            return test.PASS()
+        else:
+            return test.FAIL("Version > 1 not supported yet.")
+
+    def load_resource_data(self):
+        """Loads test data from files"""
+        result_data = dict()
```
[the remark this answers is not in the document]
Comment (jasperpeeters) [reply IS0402Test.py]: Updated in f91707e

**Comment by andrewbonney** [comment]
That's working better. One suggestion, could the DELETE tests be changed to explicitly DELETE each of the registered resources (in reverse order to how they were registered). At present it's inadvertently also performing a secondary test of 'test_28'.

**Comment by jasperpeeters** [comment]
> could the DELETE tests be changed to explicitly DELETE each of the registered resources (in reverse order to how they were registered). At present it's inadvertently also performing a secondary test of 'test_28'.

Agreed

File: IS0402Test.py [hunk 2]
```diff
@@ -816,6 +819,241 @@ def test_30(self):
         else:
             return test.FAIL("Version > 1 not supported yet.")
 
+    def test_31(self):
+        """Query API sends correct websocket event messages for SYNC, ADD, MODIFY and DELETE"""
+
+        test = Test("Query API sends correct websocket event messages for SYNC, ADD, MODIFY and DELETE")
+        api = self.apis[REG_API_KEY]
```
Comment (andrewbonney) [inline IS0402Test.py]:
```suggestion
        api = self.apis[QUERY_API_KEY]
```

File: IS0402Test.py [hunk 3]
```diff
@@ -816,6 +819,241 @@ def test_30(self):
...
+                                                            self.apis[REG_API_KEY]["version"])
+
+                result_data[resource] = resource_json
+        return result_data
+
+    def load_subscription_request_data(self):
+        """Loads subscription request data"""
+        with open("test_data/IS0402/subscriptions_request.json") as resource_data:
+            resource_json = json.load(resource_data)
+            if self.is04_reg_utils.compare_api_version(self.apis[REG_API_KEY]["version"], "v1.2") < 0:
```
Comment (andrewbonney) [inline IS0402Test.py]:
```suggestion
            if self.is04_reg_utils.compare_api_version(self.apis[QUERY_API_KEY]["version"], "v1.2") < 0:
```

File: IS0402Test.py [hunk 4]
```diff
@@ -816,6 +819,241 @@ def test_30(self):
...
+                return self.downgrade_resource("subscription", resource_json, self.apis[REG_API_KEY]["version"])
```
Comment (andrewbonney) [inline IS0402Test.py]:
```suggestion
                return self.downgrade_resource("subscription", resource_json, self.apis[QUERY_API_KEY]["version"])
```

File: IS0402Test.py [hunk 5]
```diff
@@ -816,6 +819,241 @@ def test_30(self):
...
+                    json_msg = json.loads(curr_msg)
+                    grain_data.extend(json_msg["grain"]["data"])
+
+                found_data_set = False
+                for curr_data in grain_data:
+                    pre_data = json.dumps(curr_data["pre"], sort_keys=True)
+                    sorted_resource_data = json.dumps(resource_data, sort_keys=True)
+
+                    if pre_data == sorted_resource_data:
+                        if "post_data" not in curr_data:
```
Comment (andrewbonney) [inline IS0402Test.py]:
```suggestion
                        if "post" not in curr_data:
```

File: IS0402Test.py [hunk 6]
```diff
@@ -816,6 +819,241 @@ def test_30(self):
...
+                    json_msg = json.loads(curr_msg)
+                    grain_data.extend(json_msg["grain"]["data"])
+
+                found_data_set = False
+                for curr_data in grain_data:
+                    post_data = json.dumps(curr_data["post"], sort_keys=True)
+                    sorted_resource_data = json.dumps(resource_data, sort_keys=True)
+
+                    if post_data == sorted_resource_data:
+                        if "pre_data" not in curr_data:
```
Comment (andrewbonney) [inline IS0402Test.py]:
```suggestion
                        if "pre" not in curr_data:
```

**Comment by garethsb-sony** [comment]
I've just confirmed nmos-cpp-registry passes these tests. I'll try to dive into more detail soon.

**Comment by andrewbonney** [comment]
@garethsb-sony would you like some more time to look through the change before it's merged in. It's working for me too (subject to fixing an implementation error I've just confirmed).

File: IS0402Test.py [hunk 7]
```diff
@@ -816,6 +819,243 @@ def test_30(self):
...
+
+        test = Test("Query API sends correct websocket event messages for SYNC, ADD, MODIFY and DELETE")
+        api = self.apis[QUERY_API_KEY]
+
+        if self.is04_reg_utils.compare_api_version(api["version"], "v2.0") <= 0:
+
+            # Check for clean state // delete resources if needed
+            valid, r = self.do_request("GET", "{}nodes/{}".format(self.query_url, self.test_data["node"]["id"]))
+            if not valid:
+                return test.FAIL("Query API returend an unexpected response: {}".format(r.status_code, r.text))
```
Comment (garethsb-sony) [inline IS0402Test.py]: Trivial typo = returend -> returned

File: IS0402Test.py [hunk 8]
```diff
@@ -816,6 +819,243 @@ def test_30(self):
         else:
             return test.FAIL("Version > 1 not supported yet.")
 
+    def test_31(self):
+        """Query API sends correct websocket event messages for SYNC, ADD, MODIFY and DELETE"""
```
Comment (garethsb-sony) [inline IS0402Test.py]: Should we aim for consistent terminology with the spec [WebSocket protocol](https://github.com/AMWA-TV/nmos-discovery-registration/blob/v1.2.x/docs/4.2.%20Behaviour%20-%20Querying.md#websocket-protocol), i.e. "sync" (or "unchanged"), "added", "modified" and "removed"?

File: IS0402Test.py [hunk 9]
```diff
@@ -816,6 +819,243 @@ def test_30(self):
...
+
+        test = Test("Query API sends correct websocket event messages for SYNC, ADD, MODIFY and DELETE")
+        api = self.apis[QUERY_API_KEY]
+
+        if self.is04_reg_utils.compare_api_version(api["version"], "v2.0") <= 0:
```
Comment (garethsb-sony) [inline IS0402Test.py]: ...``, "v2.0") <= 0`` means less than *or equal* to v2.0. I think we want ``<`` here.

File: IS0402Test.py [hunk 10]
```diff
@@ -816,6 +819,243 @@ def test_30(self):
...
+                            found_data_set = True
+
+                if not found_data_set:
+                    return test.FAIL("Did not found expected data set in websocket DELETE message for '{}'"
+                                     .format(resource))
+
+            # Verify if corresponding message received via Websocket: ADD
+            # Post sample data again
+            for resource in resources_to_post:
+                valid, r = self.do_request("POST", self.reg_url + "resource", data={"type": resource,
```
Comment (garethsb-sony) [inline IS0402Test.py]: It would be elegant to post the resource data with a bumped version timestamp, rather than the original one.

**Comment by jasperpeeters** [comment]
@garethsb-sony :+1:

**Pull request merged**